specs/ChainSpec/Upgradability.md

```diff
@@ -20,11 +20,18 @@ type ProtocolVersion = u32;
 
 ## Client versioning
 
-Clients should follow [semantic versioning](https://semver.org/).
+Clients should follow [semantic versioning](https://semver.org/). Our client is available both as an executable and as a library. We keep their versions in sync since executable is just a command line parser over the library.
+
 Specifically:
- - MAJOR version defines protocol releases.
- - MINOR version defines changes that are client specific but require database migration, change of config or something similar. This includes client-specific features. Client should execute migrations on start, by detecting that information on disk is produced by previous version and auto-migrate it to new one.
-  - PATCH version defines when bug fixes, which should not require migrations or protocol changes.
+ - MAJOR version defines non-backward compatible API and behavior changes. Specifically, if any kind of old client input produces different output on the new client, whether it is used an executable or a library. Whether it is being used by another client implementation, a third-party service, like bridge, wallet, or by another Rust crate, e.g. for Explorer Indexer. Examples of executable client input: RPC requests (including contract calls), incoming network messages, disk storage. Examples of executable client output: RPC responses, outgoing network messages, modifications to disk storage. Examples of library client input: messages send to publicly accessible actors. Examples of library client output: messages received from publicly accessible actors. Non-intuitive examples of when MAJOR version update is needed:
+   - Anything that requires database migration, even if client self-migrates its database;
+   - Error format of the RPC, including compilation errors. Which implies that major updates to Wasm runtime can cause version bump, even if they do not change the protocol;
+   - The way publicly accessible actors handle messages;
+   - When client changes its behavior after X's block -- delayed non-backward compatible API and behavior change, which includes most of the protocol changes.
+- MINOR version defines backward compatible API and behavior changes, this mostly applies to extensions Non-intuitive examples of when MINOR version upgrade is needed:
+   - New RPC endpoints;
+   - Additional optional storage columns, e.g. on-disk cache.
```

Comment:
But it still requires a migration to be adopted? Also I don't fully understand why database migration implies a major version change.

Reply:
> why database migration implies a major version change.

If you cannot interchangeably use the binaries, those should have different major version. If you perform a migration and unable to run the previous version of binary, they are not compatible 🤷

Reply:
I see. Makes sense

Reply:
What is the point of patch then if minor version should be fully compatible for running?
From what I read it says minor is any backward compatible change but doesn't mean that you can still run old binary - because if some feature was added it's possible you can't reverse anymore due to some changes even in installable software.

Reply:
Yes, good point. AFAIU the difference between patch and minor version could be purely advisory -- patch means the users should upgrade their node (we fixed a bug or improved performance), while minor version means user don't have to upgrade their node, but they can. Minor versions can introduce deprecation warnings to old interfaces, which might cause some user not wanting to upgrade.